src/DotNetty.Codecs/MessageToByteEncoder.cs

```diff
@@ -52,13 +53,13 @@ public override Task WriteAsync(IChannelHandlerContext context, object message)
                     return context.WriteAsync(message);
                 }
             }
-            catch (EncoderException e)
+            catch (EncoderException)
             {
-                return TaskEx.FromException(e);
+                throw;
             }
             catch (Exception ex)
             {
-                return TaskEx.FromException(new EncoderException(ex));
+                throw new EncoderException(ex);
```

Comment:
throwing synchronously in async method is a bad practice. Can we return failure in ValueTask?

Reply:
Any reason it was diverged from original netty in a first place:
https://github.com/netty/netty/blob/0692bf1b6a6764dee22132630bef7ee2f850771e/codec/src/main/java/io/netty/handler/codec/MessageToByteEncoder.java#L125

Reply:
and what happens to that exception?

Reply:
Presumable it baubles up all the way to a caller

Reply:
However ChannelHandlerContext always converts it to failed future:
https://github.com/netty/netty/blob/74f24a5c19f8f351e9c6a7a84bdd9fbcc7a07ada/transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java#L740

So probably context is assumed to be higher level exception handler here.

Reply:
so, pro: we're collecting stack trace which might be useful debugging real issue
cons: we're collecting stack trace which is costly (prohibitively so in some scenarios).
not sure what is worse here. I'd guess if we wanted to return a failed future here it would be a valuetask wrapping failed Task, right?

Reply:
main pro I see: we're close to original source code so it won't cause unexpected behavior when porting code based on this encoder/decoder infra in future.

For failed future yes, Exception -> Task -> ValueTask. Stephen Toub explicitly said that exceptions are already expensive enough and making ValueTask reference exception instance directly doesn't make sense.

Reply:
@nayato @maksimkim If we assume Exceptions are rare, it would ok to accept the perf hit.